# Problem: Predicting Airplane Delays

The goals of this notebook are:
- Process and create a dataset from downloaded ZIP files
- Exploratory data analysis (EDA)
- Establish a baseline model and improve it

## Introduction to business scenario
You work for a travel booking website that is working to improve the customer experience for flights that were delayed. The company wants to create a feature to let customers know if the flight will be delayed due to weather when the customers are booking the flight to or from the busiest airports for domestic travel in the US. 

You are tasked with solving part of this problem by leveraging machine learning to identify whether the flight will be delayed due to weather. You have been given access to the a dataset of on-time performance of domestic flights operated by large air carriers. You can use this data to train a machine learning model to predict if the flight is going to be delayed for the busiest airports.

### Dataset
The provided dataset contains scheduled and actual departure and arrival times reported by certified US air carriers that account for at least 1 percent of domestic scheduled passenger revenues. The data was collected by the Office of Airline Information, Bureau of Transportation Statistics (BTS). The dataset contains date, time, origin, destination, airline, distance, and delay status of flights for flights between 2014 and 2018.
The data are in 60 compressed files, where each file contains a CSV for the flight details in a month for the five years (from 2014 - 2018). The data can be downloaded from this [link](https://ucstaff-my.sharepoint.com/:f:/g/personal/ibrahim_radwan_canberra_edu_au/Er0nVreXmihEmtMz5qC5kVIB81-ugSusExPYdcyQTglfLg?e=bNO312). Please download the data files and place them on a relative path. Dataset(s) used in this assignment were compiled by the Office of Airline Information, Bureau of Transportation Statistics (BTS), Airline On-Time Performance Data, available with the following [link](https://www.transtats.bts.gov/Fields.asp?gnoyr_VQ=FGJ). 

# Step 1: Prepare the environment 

Use one of the labs which we have practised on with the Amazon Sagemakers where you perform the following steps:
1. Start a lab.
2. Create a notebook instance and name it "oncloudproject".
3. Increase the used memory to 25 GB from the additional configurations.
4. Open Jupyter Lab and upload this notebook into it.
5. Upload the two combined CVS files (combined_csv_v1.csv and combined_csv_v2.csv), which you created in Part A of this project.

# Step 2: Build and evaluate simple models

Write code to perform the follwoing steps:
1. Split data into training, validation and testing sets (70% - 15% - 15%).
2. Use linear learner estimator to build a classifcation model.
3. Host the model on another instance
4. Perform batch transform to evaluate the model on testing data
5. Report the performance metrics that you see better test the model performance 

Note: You are required to perform the above steps on the two combined datasets separatey and to comments on the difference.

In [1]:
import pandas as pd
import numpy as np
import os

In [2]:
df_1 = pd.read_csv("data/combined_csv_v1.csv")
df_1.head(5)

target  Distance  Quarter_2  Quarter_3  Quarter_4  Month_2  Month_3  \
0     0.0    1464.0          0          0          1        0        0   
1     1.0    1464.0          0          0          1        0        0   
2     0.0    1464.0          0          0          1        0        0   
3     0.0    1464.0          0          0          1        0        0   
4     0.0    1464.0          0          0          1        0        0   

   Month_4  Month_5  Month_6  ...  DepHourofDay_14  DepHourofDay_15  \
0        0        0        0  ...                0                0   
1        0        0        0  ...                0                0   
2        0        0        0  ...                0                0   
3        0        0        0  ...                0                0   
4        0        0        0  ...                0                0   

   DepHourofDay_16  DepHourofDay_17  DepHourofDay_18  DepHourofDay_19  \
0                0                0                0                0   
1                0                0                0                0   
2                0                0                0                0   
3                0                0                0                0   
4                0                0                0                0   

   DepHourofDay_20  DepHourofDay_21  DepHourofDay_22  DepHourofDay_23  
0                0                0                0                0  
1                0                0                0                0  
2                0                0                0                0  
3                0                0                0                0  
4                0                0                0                0  

[5 rows x 94 columns]

In [3]:
print("\nInfo for combined_csv_v1:")
print(df_1.info())


Info for combined_csv_v1:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1635590 entries, 0 to 1635589
Data columns (total 94 columns):
 #   Column                Non-Null Count    Dtype  
---  ------                --------------    -----  
 0   target                1635590 non-null  float64
 1   Distance              1635590 non-null  float64
 2   Quarter_2             1635590 non-null  int64  
 3   Quarter_3             1635590 non-null  int64  
 4   Quarter_4             1635590 non-null  int64  
 5   Month_2               1635590 non-null  int64  
 6   Month_3               1635590 non-null  int64  
 7   Month_4               1635590 non-null  int64  
 8   Month_5               1635590 non-null  int64  
 9   Month_6               1635590 non-null  int64  
 10  Month_7               1635590 non-null  int64  
 11  Month_8               1635590 non-null  int64  
 12  Month_9               1635590 non-null  int64  
 13  Month_10              1635590 non-null  int64  
 14  Month_1

In [4]:
# Check for missing values
print("\nMissing values in combined_csv_v1:")
print(df_1.isnull().sum())


Missing values in combined_csv_v1:
target             0
Distance           0
Quarter_2          0
Quarter_3          0
Quarter_4          0
                  ..
DepHourofDay_19    0
DepHourofDay_20    0
DepHourofDay_21    0
DepHourofDay_22    0
DepHourofDay_23    0
Length: 94, dtype: int64


## File combined_csv_v1

In [5]:
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

In [6]:
X_1=df_1.drop('target', axis=1)
y_1=df_1['target']

In [7]:
# Split data into training (70%), validation (15%), and testing (15%)
X_train_1, X_temp_1, y_train_1, y_temp_1 = train_test_split(X_1, y_1, test_size=0.3, random_state=42)
X_valid_1, X_test_1, y_valid_1, y_test_1 = train_test_split(X_temp_1, y_temp_1, test_size=0.5, random_state=42)

# Convert to NumPy arrays
X_train_1 = X_train_1.to_numpy()
y_train_1 = y_train_1.to_numpy()
X_valid_1 = X_valid_1.to_numpy()
y_valid_1 = y_valid_1.to_numpy()
X_test_1 = X_test_1.to_numpy()
y_test_1 = y_test_1.to_numpy()

In [8]:
# Step 4: Prepare Data for SageMaker
os.makedirs('data/train', exist_ok=True)
os.makedirs('data/validation', exist_ok=True)
os.makedirs('data/test', exist_ok=True)
os.makedirs('data/batch', exist_ok=True)

In [9]:
# Save the data as CSV files
train_data_1 = pd.DataFrame(np.column_stack((y_train_1, X_train_1)))
train_data_1.to_csv('data/train/train_1.csv', header=False, index=False)

valid_data_1 = pd.DataFrame(np.column_stack((y_valid_1, X_valid_1)))
valid_data_1.to_csv('data/validation/validation_1.csv', header=False, index=False)

test_data_1 = pd.DataFrame(np.column_stack((y_test_1, X_test_1)))
test_data_1.to_csv('data/test/test_1.csv', header=False, index=False)

In [10]:
import boto3
import sagemaker
from sagemaker import image_uris
from sagemaker import Session
from sagemaker.estimator import Estimator

sagemaker.config INFO - Not applying SDK defaults from location: /etc/xdg/sagemaker/config.yaml
sagemaker.config INFO - Not applying SDK defaults from location: /home/ec2-user/.config/sagemaker/config.yaml


In [11]:
# Set up SageMaker
sess = sagemaker.Session()
bucket = sess.default_bucket()
prefix = 'sagemaker/linear-v1'
region = boto3.Session().region_name

# Get the Linear Learner container image
linear_learner_container = image_uris.retrieve("linear-learner", region)

In [12]:
# Upload the training and validation data to S3
s3 = boto3.client('s3')

train_path_1 = sess.upload_data(path='data/train/train_1.csv', key_prefix=f'{prefix}/input/training')
valid_path_1 = sess.upload_data(path='data/validation/validation_1.csv', key_prefix=f'{prefix}/input/validation')
test_path_1 = sess.upload_data(path='data/test/test_1.csv', key_prefix=f'{prefix}/input/test')

In [13]:
# Print S3 paths
print(f'Training data uploaded to: {train_path_1}')
print(f'Validation data uploaded to: {valid_path_1}')
print(f'Test data uploaded to: {test_path_1}')

Training data uploaded to: s3://sagemaker-us-east-1-807330624080/sagemaker/linear-v1/input/training/train_1.csv
Validation data uploaded to: s3://sagemaker-us-east-1-807330624080/sagemaker/linear-v1/input/validation/validation_1.csv
Test data uploaded to: s3://sagemaker-us-east-1-807330624080/sagemaker/linear-v1/input/test/test_1.csv


In [14]:
# Define the Linear Learner Estimator
linear_estimator = Estimator(
    linear_learner_container,
    role=sagemaker.get_execution_role(),
    instance_count=1,
    instance_type='ml.m5.xlarge',
    output_path='s3://{}/{}/output'.format(bucket,prefix))

# Setting mini_batch_size to 100 since the dataset is large
linear_estimator.set_hyperparameters(predictor_type='binary_classifier',mini_batch_size=100, epochs=10)


In [15]:
# Define the data channels with the correct content type
training_data_channel = sagemaker.TrainingInput(s3_data=train_path_1, content_type='text/csv')
validation_data_channel = sagemaker.TrainingInput(s3_data=valid_path_1, content_type='text/csv')


# Training Phase
### Fitting the model

In [16]:
# Fit the model with the correct training and validation data channels
linear_estimator.fit({'train': training_data_channel, 'validation': validation_data_channel})

INFO:sagemaker:Creating training-job with name: linear-learner-2024-10-30-11-38-01-854


2024-10-30 11:38:03 Starting - Starting the training job...
2024-10-30 11:38:18 Starting - Preparing the instances for training...
2024-10-30 11:38:49 Downloading - Downloading input data...
2024-10-30 11:39:24 Downloading - Downloading the training image......
2024-10-30 11:40:30 Training - Training image download completed. Training in progress..Docker entrypoint called with argument(s): train
Running default environment configuration script
[10/30/2024 11:40:41 INFO 139830609086272] Reading default configuration from /opt/amazon/lib/python3.8/site-packages/algorithm/resources/default-input.json: {'mini_batch_size': '1000', 'epochs': '15', 'feature_dim': 'auto', 'use_bias': 'true', 'binary_classifier_model_selection_criteria': 'accuracy', 'f_beta': '1.0', 'target_recall': '0.8', 'target_precision': '0.8', 'num_models': 'auto', 'num_calibration_samples': '10000000', 'init_method': 'uniform', 'init_scale': '0.07', 'init_sigma': '0.01', 'init_bias': '0.0', 'optimizer': 'auto', 'loss': '

# Batch Transform Deployment

In [26]:
transformer = linear_estimator.transformer(instance_count=1, 
                                           instance_type="ml.m4.xlarge", 
                                           assemble_with="Line", 
                                           output_path=f"s3://{bucket}/{prefix}/batch_output")

INFO:sagemaker:Creating model with name: linear-learner-2024-10-30-12-54-34-537


In [27]:
test_df_1 = pd.read_csv("data/test/test_1.csv")
test_1_batch = test_df_1[test_df_1.columns[1:]]
test_1_batch.to_csv("data/batch/batch_input_1.csv", index = False, header = False)

In [28]:
batch_test_path = sess.upload_data(path="data/batch/batch_input_1.csv", key_prefix=prefix + "/batch_input")

In [29]:
transformer.transform(batch_test_path, content_type = "text/csv", split_type="Line")
transformer.wait()

INFO:sagemaker:Creating transform job with name: linear-learner-2024-10-30-12-54-47-847


................................................Docker entrypoint called with argument(s): serve
Running default environment configuration script
[10/30/2024 13:02:50 INFO 140061952366400] Memory profiler is not enabled by the environment variable ENABLE_PROFILER.
/opt/amazon/lib/python3.8/site-packages/mxnet/model.py:97: SyntaxWarning: "is" with a literal. Did you mean "=="?
  if num_device is 1 and 'dist' not in kvstore:
/opt/amazon/lib/python3.8/site-packages/scipy/optimize/_shgo.py:495: SyntaxWarning: "is" with a literal. Did you mean "=="?
  if cons['type'] is 'ineq':
/opt/amazon/lib/python3.8/site-packages/scipy/optimize/_shgo.py:743: SyntaxWarning: "is not" with a literal. Did you mean "!="?
  if len(self.X_min) is not 0:
[10/30/2024 13:02:54 WARNING 140061952366400] Loggers have already been setup.
[10/30/2024 13:02:54 INFO 140061952366400] loaded entry point class algorithm.serve.server_config:config_api
[10/30/2024 13:02:54 INFO 140061952366400] loading entry points
[10/30/20

## Testing performance

In [36]:
import io

y_file = boto3.client("s3").get_object(Bucket = bucket, Key = f"{prefix}/batch_output/batch_input_1.csv.out")
y_pred = pd.read_csv(io.BytesIO(y_file["Body"].read()), header = None, names = ["Predicted"])

In [37]:
y_pred["target"] = y_pred.index
y_pred

Predicted                target
{"predicted_label":0  score:0.086572475731372}  {"predicted_label":0
{"predicted_label":0  score:0.186020836234092}  {"predicted_label":0
{"predicted_label":0  score:0.085835359990596}  {"predicted_label":0
{"predicted_label":0   score:0.17403869330883}  {"predicted_label":0
{"predicted_label":0  score:0.057204093784093}  {"predicted_label":0
...                                        ...                   ...
{"predicted_label":0  score:0.192262724041938}  {"predicted_label":0
{"predicted_label":0  score:0.199697285890579}  {"predicted_label":0
{"predicted_label":0  score:0.333425402641296}  {"predicted_label":0
{"predicted_label":0  score:0.299674093723297}  {"predicted_label":0
{"predicted_label":0  score:0.104805827140808}  {"predicted_label":0

[245338 rows x 2 columns]

In [38]:
y_pred = y_pred['target'].apply(lambda x: 1 if x == 1 else 0)

y_true = test_df_1.iloc[:, 0]
accuracy = accuracy_score(y_true, y_pred)

/home/ec2-user/anaconda3/envs/python3/lib/python3.10/site-packages/sklearn/metrics/_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 due to no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))


In [39]:
print(f"Accuracy: {accuracy:.4f}")

Accuracy: 0.7908


Even though the accuracy of both models is the same at 0.7908, the precision and recall of model 2 are higher than model 1. By adding two key features such as holiday dates and weather data, enhance the model performance and increase the recall value. This highlights the relevant feature engineering in improving model performance.

# Step 3: Build and evaluate ensembe models

Write code to perform the follwoing steps:
1. Split data into training, validation and testing sets (70% - 15% - 15%).
2. Use xgboost estimator to build a classifcation model.
3. Host the model on another instance
4. Perform batch transform to evaluate the model on testing data
5. Report the performance metrics that you see better test the model performance 
6. write down your observation on the difference between the performance of using the simple and ensemble models.
Note: You are required to perform the above steps on the two combined datasets separatey.

In [19]:
import pandas as pd

df_1 = pd.read_csv('data/combined_csv_v1.csv')

In [3]:
import boto3
import sagemaker
from sagemaker import image_uris
from sagemaker import Session
from sagemaker.estimator import Estimator

sagemaker.config INFO - Not applying SDK defaults from location: /etc/xdg/sagemaker/config.yaml
sagemaker.config INFO - Not applying SDK defaults from location: /home/ec2-user/.config/sagemaker/config.yaml


In [4]:
sess = sagemaker.Session()
bucket = sess.default_bucket()

region = boto3.Session().region_name
container = image_uris.retrieve('xgboost', region,version='latest')

In [20]:
#from sklearn.model_selection import train_test_split
train, validation = train_test_split(df_1, test_size=0.3)
validation, test = train_test_split(validation, test_size = 0.5)

In [23]:
train.to_csv("data/train/train.csv", index = False, header = False)
validation.to_csv("data/validation/validation.csv", index = False, header = False)

In [24]:
test_X = test.drop(test.columns[0], axis=1)
test_y = test[test.columns[0]]
test_X.to_csv("data/test/test.csv", index = False, header = False)

In [25]:
prefix = 'sagemaker/xgboost-v1'
train_path = sess.upload_data(path="data/train/train.csv", key_prefix=prefix + "input/training")
valid_path = sess.upload_data(path="data/validation/validation.csv", key_prefix=prefix + "input/validation")
test_X_path = sess.upload_data(path="data/test/test.csv", key_prefix=prefix + "input/test")

In [30]:
xgboost_estimator = Estimator(container, 
                              role=sagemaker.get_execution_role(), 
                              instance_count=1, 
                              instance_type='ml.m4.xlarge',
                            output_path='s3://{}/{}/output'.format(bucket,prefix))

In [31]:
xgboost_estimator.set_hyperparameters(objective='binary:logistic', num_round=20)

In [32]:
training_data_channel = sagemaker.TrainingInput(s3_data=train_path, content_type='text/csv')
validation_data_channel = sagemaker.TrainingInput(s3_data=valid_path, content_type='text/csv')

In [33]:
xgboost_estimator.fit({'train': training_data_channel,'validation': validation_data_channel})

INFO:sagemaker:Creating training-job with name: xgboost-2024-10-30-22-35-43-916


2024-10-30 22:35:45 Starting - Starting the training job...
2024-10-30 22:35:58 Starting - Preparing the instances for training...
2024-10-30 22:36:23 Downloading - Downloading input data...
2024-10-30 22:37:03 Downloading - Downloading the training image...
2024-10-30 22:37:49 Training - Training image download completed. Training in progress...Arguments: train
[2024-10-30:22:37:59:INFO] Running standalone xgboost training.
[2024-10-30:22:37:59:INFO] File size need to be processed in the node: 257.72mb. Available memory size in the node: 8457.41mb
[2024-10-30:22:37:59:INFO] Determined delimiter of CSV input is ','
[22:37:59] S3DistributionType set as FullyReplicated
[22:38:02] 1144913x93 matrix with 106476909 entries loaded from /opt/ml/input/data/train?format=csv&label_column=0&delimiter=,
[2024-10-30:22:38:02:INFO] Determined delimiter of CSV input is ','
[22:38:02] S3DistributionType set as FullyReplicated
[22:38:03] 245338x93 matrix with 22816434 entries loaded from /opt/ml/input/

In [34]:
xgboost_transformer = xgboost_estimator.transformer(
    instance_count=1,
    instance_type="ml.m4.xlarge",
    strategy="MultiRecord",
    assemble_with="Line",
    output_path="s3://{}/{}/output".format(bucket, prefix),
)

INFO:sagemaker:Creating model with name: xgboost-2024-10-30-22-40-02-611


In [35]:
xgboost_transformer.transform(test_X_path, content_type="text/csv", split_type="Line")
xgboost_transformer.wait()

INFO:sagemaker:Creating transform job with name: xgboost-2024-10-30-22-40-03-751


...........................................Arguments: serve
[2024-10-30 22:47:15 +0000] [1] [INFO] Starting gunicorn 19.9.0
[2024-10-30 22:47:15 +0000] [1] [INFO] Listening at: http://0.0.0.0:8080 (1)
[2024-10-30 22:47:15 +0000] [1] [INFO] Using worker: gevent
[2024-10-30 22:47:15 +0000] [21] [INFO] Booting worker with pid: 21
[2024-10-30 22:47:15 +0000] [22] [INFO] Booting worker with pid: 22
[2024-10-30 22:47:15 +0000] [23] [INFO] Booting worker with pid: 23
/opt/amazon/lib/python3.7/site-packages/gunicorn/workers/ggevent.py:65: MonkeyPatchWarning: Monkey-patching ssl after ssl has already been imported may lead to errors, including RecursionError on Python 3.6. It may also silently lead to incorrect behaviour on Python 3.7. Please monkey-patch earlier. See https://github.com/gevent/gevent/issues/1016. Modules that had direct imports (NOT patched): ['urllib3.util (/opt/amazon/lib/python3.7/site-packages/urllib3/util/__init__.py)', 'urllib3.util.ssl_ (/opt/amazon/lib/python3.7/site-pa

In [37]:
s3 = boto3.client("s3")
response = s3.list_objects_v2(Bucket=bucket, Prefix=f"{prefix}/output/")
for obj in response.get('Contents', []):
    print(obj['Key'])

sagemaker/xgboost-v1/output/test.csv.out
sagemaker/xgboost-v1/output/xgboost-2024-10-30-22-15-35-191/debug-output/training_job_end.ts
sagemaker/xgboost-v1/output/xgboost-2024-10-30-22-15-35-191/output/model.tar.gz
sagemaker/xgboost-v1/output/xgboost-2024-10-30-22-15-35-191/profiler-output/framework/training_job_end.ts
sagemaker/xgboost-v1/output/xgboost-2024-10-30-22-15-35-191/profiler-output/system/incremental/2024103022/1730326560.algo-1.json
sagemaker/xgboost-v1/output/xgboost-2024-10-30-22-15-35-191/profiler-output/system/incremental/2024103022/1730326620.algo-1.json
sagemaker/xgboost-v1/output/xgboost-2024-10-30-22-15-35-191/profiler-output/system/incremental/2024103022/1730326680.algo-1.json
sagemaker/xgboost-v1/output/xgboost-2024-10-30-22-15-35-191/profiler-output/system/incremental/2024103022/1730326740.algo-1.json
sagemaker/xgboost-v1/output/xgboost-2024-10-30-22-15-35-191/profiler-output/system/training_job_end.ts
sagemaker/xgboost-v1/output/xgboost-2024-10-30-22-31-30-113/d

In [39]:
import io
y_file = boto3.client("s3").get_object(Bucket = bucket, Key = 'sagemaker/xgboost-v1/output/test.csv.out')
y_pred = pd.read_csv(io.BytesIO(y_file["Body"].read()), header = None, names = ["Predicted"])

In [40]:
y_pred["actual"] = y_pred["Predicted"].apply(lambda x : 1 if x > 0.5 else 0)
y_pred

Predicted  actual
0        0.279765       0
1        0.332670       0
2        0.163057       0
3        0.170161       0
4        0.202313       0
...           ...     ...
245334   0.302841       0
245335   0.225716       0
245336   0.149098       0
245337   0.209781       0
245338   0.345487       0

[245339 rows x 2 columns]

In [42]:
from sklearn.metrics import accuracy_score
accuracy = accuracy_score(test_y, y_pred["actual"])
print("Accuracy:", accuracy)

Accuracy: 0.790151586172602


When comparing the performance of the simple linear model and the ensemble model (XGBoost), both yielded similar accuracy levels, with the linear model achieving an accuracy of 0.7908 and XGBoost slightly lower at approximately 0.7902. The difference between the two methods is not significant, suggesting that both models are effective for this dataset. Additionally, XGBoost was less time-consuming during training compared to the linear model, even with the same dataset size and features. This efficiency, combined with XGBoost’s ability to capture complex relationships, makes it a valuable alternative despite the minor difference in accuracy.